### PR TITLE
8317763: Follow-up to AVX512 intrinsics for Arrays.sort() PR

### PR DESCRIPTION
**Pull request by vamsi-parasa**
The goal of this PR is to address the follow-up comments to the SIMD accelerated sort PR (#14227) which implemented AVX512 intrinsics for Arrays.sort() methods.
The proposed changes are:

1) Restriction of the AVX512 sort acceleration to only Intel CPUs. A performance regression (due to micro-architectural differences) was reported for AMD Zen4 CPUs in the comments section of PR.
2) Addressing the build failure due to a bug in GCC 12 (which was fixed in version 12.3.1). The details of the bug are at: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105593
3) Minor changes in Javadoc strings

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8317763](https://bugs.openjdk.org/browse/JDK-8317763): Follow-up to AVX512 intrinsics for Arrays.sort() PR (**Task** - P4)

### Reviewers
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**) ⚠️ Review applies to [1138dc38](https://git.openjdk.org/jdk/pull/16124/files/1138dc3883123b0331e535ca651505588463c60d)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**) ⚠️ Review applies to [1de07fb7](https://git.openjdk.org/jdk/pull/16124/files/1de07fb7cd8293481813a2c2b79aa5aac07b561a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16124/head:pull/16124` \
`$ git checkout pull/16124`

Update a local copy of the PR: \
`$ git checkout pull/16124` \
`$ git pull https://git.openjdk.org/jdk.git pull/16124/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16124`

View PR using the GUI difftool: \
`$ git pr show -t 16124`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16124.diff">https://git.openjdk.org/jdk/pull/16124.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16124#issuecomment-1755855766)

### CODE REVIEW

**Review by jatin-bhateja — APPROVED** [review]
LGTM

**Review by vnkozlov — COMMENTED** [review]
What is change for "Addressing the build failure due to a bug in GCC 12"?

**Comment by vamsi-parasa** [comment]
> What is change for "Addressing the build failure due to a bug in GCC 12"?

Hello Vladimir,

The change for addressing the build failure will be pushed shortly. 

Thanks,
Vamsi

**Comment by R1chterScale** [comment]
Forgive me, I might be missing something very obvious, but is there any particular reason to entirely disable the SIMD accelerated sort on Zen 4 rather than having an alternate code path for Zen 4 where it has the `compressstoreu` instructions split up into separate `compress` and `storeu` instructions so that Zen 4 platforms can still benefit from a decent degree of performance uplift from AVX512 acceleration of sort?

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (vamsi-parasa) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Please see this fixed in the latest commit pushed.

**Comment by vamsi-parasa** [comment]
Hello Vladimir (@vnkozlov) and Erik (@erikj79),

Please see the latest commit which has the `#pragma` based workaround to address the build failure caused by GCC 12.

Thanks,
Vamsi

**Comment by magicus** [comment]
In #14227, you inadvertently added an extra space at line 230 in make/modules/java.base/Lib.gmk
(https://github.com/openjdk/jdk/pull/14227/files#diff-c2e113e4b2661697750fd5e6dcc0908fa98563ccfb3801c8b0e3a70174041b81).

Can you please revert it here?

**Comment by vamsi-parasa** [comment]
> In #14227, you inadvertently added an extra space at line 230 in make/modules/java.base/Lib.gmk 
Hi Magnus (@magicus), please see the extra space fixed in the latest commit.

Thanks,
Vamsi

**Comment by iaroslavski** [comment]
Hi @vamsi-parasa ,

Methods `partitionDualPivot` and `partitionSinglePivot` are annotated by @ForceInline, but
`mixedInsertionSort` and `insertionSort` are not. These all 4 methods are passed as parameter 
to methods `private static <A> int[] partition(Class<?> elemType, A array, ...`
and `private static <A> void sort(Class<?> elemType, A array, ...`.

Is ok that `partitionDualPivot`, `partitionSinglePivot` and `mixedInsertionSort`, `insertionSort` are annotated differently?

Should we add @ForceInline to `mixedInsertionSort`, `insertionSort` or
remove @ForceInline from `partitionDualPivot` and `partitionSinglePivot`?

**Comment by vnkozlov** [comment]
> Is ok that partitionDualPivot, partitionSinglePivot and mixedInsertionSort, insertionSort are annotated differently?

Good question. Someone familiar with this Java code should answer.

Note,  **@forceinline** annotation is used by C2 JIT compiler when it decide to inline. One criteria is the bytecode size (we don't want inline huge method into another compilation). Default value is 35. With **@forceinline** the size will be **1** - most likely guarantee inlining.
But there is drawback, if java method is not "hot" (not called frequently) the inlined code may lead to not optimal code generation.
Use it carefully.

**Comment by vnkozlov** [comment]
Also @forceinline in these changes only works for case when new intrinsics are not used.
I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics.
That will tell us which @forceinline annotations are needed.

**Comment by iaroslavski** [comment]
> Also @forceinline in these changes only works for case when new intrinsics are not used. I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics. That will tell us which @forceinline annotations are needed.

@vnkozlov Thank you for explanation, I will benchmark cases with and without @forceinline and provide results very soon.

**Comment by iaroslavski** [comment]
@vnkozlov 
Are there any options to be sure that C2 JIT compiler is used during tests?

**Comment by iaroslavski** [comment]
Hi @vamsi-parasa,

Both methods mixedInsertionSort and insertionSort are covered by intrinsics.
But insertionSort is run on leftmnost (one) part only and on small ( < MAX_INSERTION_SORT_SIZE = 44) arrays.
Do we actually need to use intrinsics for it?

To have clear picture could you please run benchmarking to compare both cases:
current implementation and implementation with Java insertionSort only?

see changes `sort(int.class, a, Unsafe.ARRAY_INT_BASE_OFFSET, low, high, DualPivotQuicksort::insertionSort);`
->
`insertionSort(a, low, high);`

**Comment by theRealAph** [comment]
> Forgive me, I might be missing something very obvious, but is there any particular reason to entirely disable the SIMD accelerated sort on Zen 4 rather than having an alternate code path for Zen 4 where it has the `compressstoreu` instructions split up into separate `compress` and `storeu` instructions so that Zen 4 platforms can still benefit from a decent degree of performance uplift from AVX512 acceleration of sort?

I don't think you're missing anything. This should be done, rather than disabling the intrinsic.

**Review by magicus — APPROVED** [review]
Build changes look good. Thanks for fixing this!

/label +build
/reviewers 2

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Comment by vamsi-parasa** [comment]
> Also @forceinline in these changes only works for case when new intrinsics are not used. I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics. That will tell us which @forceinline annotations are needed.

Added `@ForceInline` annotations to `insertionSort` and `mixedInsertionSort` as it is helping arrays of small sizes when intrinsics are disabled.

Thanks,
Vamsi

**Comment by sviswa7** [comment]
@vnkozlov Please advice if we can integrate this PR or if you would like to run some tests first.

**Comment by vnkozlov** [comment]
@R1chterScale or someone can suggest patch for Zen 4 and test it so we can include it into these changes?
Otherwise we will file separate followup RFE as Sandhya suggested?

And yes, I need to run testing before approval but after we decide what to do with Zen 4 in these changes.

**Comment by R1chterScale** [comment]
Sorry, it's probably out of my capabilities to make something of acceptable quality. All I know is that it would be useful to reference the unmerged Zen 4 patch for the base implementation of x86-simd-sort

**Comment by vnkozlov** [comment]
Okay. I will start testing for current changes. @sviswa7 please file RFE for Zen 4. If we get patch for it we do followup changes in that RFE.

**Comment by iaroslavski** [comment]
> > Also @forceinline in these changes only works for case when new intrinsics are not used. I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics. That will tell us which @forceinline annotations are needed.
> 
> Added `@ForceInline` annotations to `insertionSort` and `mixedInsertionSort` as it is helping arrays of small sizes when intrinsics are disabled.
> 
> Thanks, Vamsi

@vamsi-parasa 
Please revert changes (adding @ForceInline to insertionSort and mixedInsertionSort) - I checked: initinal version works faster.

**Comment by vnkozlov** [comment]
> @vnkozlov
> Are there any options to be sure that C2 JIT compiler is used during tests?

May be you already figure out that. Use `-XX:+PrintCompilation -XX:+UnlockDiagnosticVMOptions -XX:+PrintInlining -XX:-TieredCompilation` flags (Tiered is off to use only C2) and look for something like next
```
     82   20             java.util.Random::next (47 bytes)
                            @ 3   java.util.Random::next (47 bytes)   inline (hot)
                              @ 8   java.util.concurrent.atomic.AtomicLong::get (5 bytes)   accessor
                              @ 32   java.util.concurrent.atomic.AtomicLong::compareAndSet (13 bytes)   inline (hot)
                                @ 9   jdk.internal.misc.Unsafe::compareAndSetLong (0 bytes)   (intrinsic)
```

**Comment by vamsi-parasa** [comment]
> > > Also @forceinline in these changes only works for case when new intrinsics are not used. I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics. That will tell us which @forceinline annotations are needed.
> > 
> > 
> > Added `@ForceInline` annotations to `insertionSort` and `mixedInsertionSort` as it is helping arrays of small sizes when intrinsics are disabled.
> > Thanks, Vamsi
> 
> @vamsi-parasa Please revert changes (adding @forceinline to insertionSort and mixedInsertionSort) - I checked: initinal version works faster.

@iaroslavski 
Vladimir, have you used `-XX:+UnlockDiagnosticVMOptions -XX:DisableIntrinsic=_arraySort,_arrayPartition` to disable the intrinsics?
Also, it helps to have one warmup iteration of 30 secs to reduce run-to-run variance.

**Comment by iaroslavski** [comment]
> > > > Also @forceinline in these changes only works for case when new intrinsics are not used. I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics. That will tell us which @forceinline annotations are needed.
> > > 
> > > 
> > > Added `@ForceInline` annotations to `insertionSort` and `mixedInsertionSort` as it is helping arrays of small sizes when intrinsics are disabled.
> > > Thanks, Vamsi
> > 
> > 
> > @vamsi-parasa Please revert changes (adding @forceinline to insertionSort and mixedInsertionSort) - I checked: initinal version works faster.
> 
> @iaroslavski Vladimir, have you used `-XX:+UnlockDiagnosticVMOptions -XX:DisableIntrinsic=_arraySort,_arrayPartition` to disable the intrinsics? Also, it helps to have one warmup iteration of 30 secs to reduce run-to-run variance.

@vamsi-parasa 
Yes, I tried with the disabled mentioned intrinsics.

**Comment by vamsi-parasa** [comment]
> To have clear picture could you please run benchmarking to compare both cases: current implementation and implementation with Java insertionSort only?
> 
> see changes `sort(int.class, a, Unsafe.ARRAY_INT_BASE_OFFSET, low, high, DualPivotQuicksort::insertionSort);` -> `insertionSort(a, low, high);`

@iaroslavski 
Yes, the intrinsic is needed. 
I remember, the same question was asked in the original PR and data was provided in the original PR.
Even for a small size=30, the AVX512 sort intrinsic is about 3x faster as seen below:

File: make/modules/java.base/Lib.gmk
Comment (erikj79) [inline make/modules/java.base/Lib.gmk:230]: If you are going to fix the indentation here, please do it to follow the code conventions for the build system:
https://openjdk.org/groups/build/doc/code-conventions.html
```
2. If a line must be broken, use four spaces for indentation.
18. Avoid padding internally in a line with spaces to try to align some feature into columns with surrounding lines.
```

**Comment by iaroslavski** [comment]
> > > > > Also @forceinline in these changes only works for case when new intrinsics are not used. I would suggest to adapt/update JMH benchmark to cover all cases and see effect @forceinline without intrinsics. That will tell us which @forceinline annotations are needed.
> > > > 
> > > > 
> > > > Added `@ForceInline` annotations to `insertionSort` and `mixedInsertionSort` as it is helping arrays of small sizes when intrinsics are disabled.
> > > > Thanks, Vamsi
> > > 
> > > @vamsi-parasa Please revert changes (adding @forceinline to insertionSort and mixedInsertionSort) - I checked: initinal version works faster.
> > 
> > @iaroslavski Vladimir, have you used `-XX:+UnlockDiagnosticVMOptions -XX:DisableIntrinsic=_arraySort,_arrayPartition` to disable the intrinsics? Also, it helps to have one warmup iteration of 30 secs to reduce run-to-run variance.
> 
> @vamsi-parasa Yes, I tried with the disabled mentioned intrinsics.

@vamsi-parasa, and no profit with @forceinline on insertionSort and mixedInsertionSort.
I suggest removing @forceinline from insertionSort and mixedInsertionSort.

**Comment by vamsi-parasa** [comment]
> @vamsi-parasa, and no profit with @forceinline on insertionSort and mixedInsertionSort. I suggest removing @forceinline from insertionSort and mixedInsertionSort.

@iaroslavski,  Please see the change reverted.

File: make/modules/java.base/Lib.gmk
Comment (vamsi-parasa) [reply make/modules/java.base/Lib.gmk:230]: Hi Erik (@erikj79), 
BUILD_LIBFALLBACKLINKER is from different PR (#13079). If I understand correctly, for LIB_SIMD_SORT, are you suggesting that we don't pad the lines with spaces to align features into columns and instead just use 4 spaces for the indentation after the line break?

**Comment by sviswa7** [comment]
> Okay. I will start testing for current changes. @sviswa7 please file RFE for Zen 4. If we get patch for it we do followup changes in that RFE.

Thanks a lot Vladimir. RFE filed: https://bugs.openjdk.org/browse/JDK-8317976.

**Comment by vnkozlov** [comment]
I started testing for v04 assuming if you make make file changes they will be cosmetic.

File: make/modules/java.base/Lib.gmk
Comment (erikj79) [reply make/modules/java.base/Lib.gmk:230]: I see now that this is an unrelated change. In that case please avoid changing whitespace in unrelated files for this PR.

Comment (sviswa7) [reply make/modules/java.base/Lib.gmk:230]: @erikj79 This space was inadvertently added as part of (https://github.com/openjdk/jdk/pull/14227) and @magicus requested to remove it in this follow-up PR.

**Comment by iwanowww** [comment]
Proposed patch has one disadvantage: there's no way to override ergonomics decisions on AMD CPUs and forcibly enable the intrinsic without rebuilding the JVM. 

For many other intrinsics there are flags which enable finer grained control over JVM behavior (e.g., `UseVectorizedHashCodeIntrinsic`). There were some complaints that there are way to many individual flags to control intrinsics, so now there's `-XX:ControlIntrinsic=`, but as of now ergonomics doesn't rely on it to control what intrinsics are enabled/disabled.

It makes sense to let `-XX:ControlIntrinsic=` overrule `VM_Version::is_intel()` check and enable the intrinsics when `AVX512DQ` is supported.

**Comment by sviswa7** [comment]
> Proposed patch has one disadvantage: there's no way to override ergonomics decisions on AMD CPUs and forcibly enable the intrinsic without rebuilding the JVM.
> 
> For many other intrinsics there are flags which enable finer grained control over JVM behavior (e.g., `UseVectorizedHashCodeIntrinsic`). There were some complaints that there are way to many individual flags to control intrinsics, so now there's `-XX:ControlIntrinsic=`, but as of now ergonomics doesn't rely on it to control what intrinsics are enabled/disabled.
> 
> It makes sense to let `-XX:ControlIntrinsic=` overrule `VM_Version::is_intel()` check and enable the intrinsics when `AVX512DQ` is supported.

May be it could be done as part of https://bugs.openjdk.org/browse/JDK-8317976.

**Comment by vnkozlov** [comment]
> > It makes sense to let `-XX:ControlIntrinsic=` overrule `VM_Version::is_intel()` check and enable the intrinsics when `AVX512DQ` is supported.
> 
> May be it could be done as part of https://bugs.openjdk.org/browse/JDK-8317976.

It is not easy to do. You need to make sure that you not enable intrinsic by `-XX:ControlIntrinsic=` when it is not supported by hardware. I would say it is separate (second) RFE because we need to test it separately from current Zen 4 issue.

Note, we are not introducing regression to AMD with these changes - code is reverted back to before previous changes [14227](https://github.com/openjdk/jdk/pull/14227). We have some  time to propose patch.

**Comment by iwanowww** [comment]
FTR I'm perfectly fine with handling it separately.

> It is not easy to do. You need to make sure that you not enable intrinsic by -XX:ControlIntrinsic=

There's `VM_Version::is_intrinsic_supported()` introduced specifically for such type of platform-specific checks. `vmIntrinsics::is_intrinsic_available()` takes into account both `ControlIntrinsic`/`DisableIntrinsic` and  `VM_Version::is_intrinsic_supported()`.

**Comment by DanielThomas** [comment]
At least on Saphire Rapids the [emulated suggested here](https://github.com/natmaurice/x86-simd-sort/commit/41d03b2d8f3b62a2ee6a3a97a8da7f193a407026) only imposes a 6% penalty for `intSort`, while also mitigating the performance issue on Zen 4.

```
Configuration summary:
* Name:           linux-x86_64-server-release
* Debug level:    release
* HS debug level: product
* JVM variants:   server
* JVM features:   server: 'cds compiler1 compiler2 epsilongc g1gc jfr jni-check jvmci jvmti management parallelgc serialgc services shenandoahgc vm-structs zgc'
* OpenJDK target: OS: linux, CPU architecture: x86, address length: 64
* Version string: 22-internal-adhoc.nfsuper.jdk (22-internal)
* Source date:    1697078366 (2023-10-12T02:39:26Z)

Tools summary:
* Boot JDK:       openjdk version "21" 2023-09-19 OpenJDK Runtime Environment Zulu21.28+86-SA (build 21+35) OpenJDK 64-Bit Server VM Zulu21.28+86-SA (build 21+35, mixed mode, sharing) (at /usr/lib/jvm/zulu-21-amd64)
* Toolchain:      gcc (GNU Compiler Collection)
* C Compiler:     Version 11.4.0 (at /usr/bin/gcc)
* C++ Compiler:   Version 11.4.0 (at /usr/bin/g++)
```

https://github.com/openjdk/jdk/compare/master...DanielThomas:jdk:dannyt/emulate-compressstoreu?expand=1

## Intel(R) Xeon(R) Platinum 8488C - Current

```
Benchmark            (size)  Mode  Cnt     Score     Error  Units
ArraysSort.intSort       10  avgt    3     0.043 ?   0.006  us/op
ArraysSort.intSort       25  avgt    3     0.082 ?   0.002  us/op
ArraysSort.intSort       50  avgt    3     0.205 ?   0.022  us/op
ArraysSort.intSort       75  avgt    3     0.394 ?   0.048  us/op
ArraysSort.intSort      100  avgt    3     0.625 ?   0.003  us/op
ArraysSort.intSort     1000  avgt    3     5.759 ?   1.111  us/op
ArraysSort.intSort    10000  avgt    3    51.680 ?   3.568  us/op
ArraysSort.intSort   100000  avgt    3   777.339 ?  25.809  us/op
ArraysSort.intSort  1000000  avgt    3  8848.261 ? 954.475  us/op
```

## Intel(R) Xeon(R) Platinum 8488C - Emulated

```
Benchmark            (size)  Mode  Cnt     Score     Error  Units
ArraysSort.intSort       10  avgt    3     0.046 ?   0.002  us/op
ArraysSort.intSort       25  avgt    3     0.083 ?   0.004  us/op
ArraysSort.intSort       50  avgt    3     0.214 ?   0.022  us/op
ArraysSort.intSort       75  avgt    3     0.411 ?   0.038  us/op
ArraysSort.intSort      100  avgt    3     0.658 ?   0.022  us/op
ArraysSort.intSort     1000  avgt    3     6.411 ?   0.497  us/op
ArraysSort.intSort    10000  avgt    3    55.996 ?   3.155  us/op
ArraysSort.intSort   100000  avgt    3   822.805 ?  40.223  us/op
ArraysSort.intSort  1000000  avgt    3  9487.974 ? 216.146  us/op
```

## Intel(R) Xeon(R) Platinum 8488C - Baseline

```
Benchmark            (size)  Mode  Cnt      Score      Error  Units
ArraysSort.intSort       10  avgt    3      0.047 ?    0.006  us/op
ArraysSort.intSort       25  avgt    3      0.099 ?    0.022  us/op
ArraysSort.intSort       50  avgt    3      0.249 ?    0.024  us/op
ArraysSort.intSort       75  avgt    3      0.438 ?    0.046  us/op
ArraysSort.intSort      100  avgt    3      0.590 ?    0.079  us/op
ArraysSort.intSort     1000  avgt    3      8.384 ?    1.852  us/op
ArraysSort.intSort    10000  avgt    3    435.589 ?   23.647  us/op
ArraysSort.intSort   100000  avgt    3   5380.658 ?  491.435  us/op
ArraysSort.intSort  1000000  avgt    3  63857.189 ? 2746.106  us/op
```

## AMD EPYC 9R14 - Emulated

```
$ make test TEST="micro:java.lang.ArraysSort.intSort"

Benchmark            (size)  Mode  Cnt     Score     Error  Units
ArraysSort.intSort       10  avgt    3     0.032 ?   0.001  us/op
ArraysSort.intSort       25  avgt    3     0.067 ?   0.002  us/op
ArraysSort.intSort       50  avgt    3     0.196 ?   0.002  us/op
ArraysSort.intSort       75  avgt    3     0.429 ?   0.046  us/op
ArraysSort.intSort      100  avgt    3     0.614 ?   0.025  us/op
ArraysSort.intSort     1000  avgt    3     6.500 ?   0.084  us/op
ArraysSort.intSort    10000  avgt    3    55.620 ?   0.943  us/op
ArraysSort.intSort   100000  avgt    3   669.347 ?  75.432  us/op
ArraysSort.intSort  1000000  avgt    3  9459.001 ? 201.298  us/op
Finished running test 'micro:java.lang.ArraysSort.intSort'
```

## AMD EPYC 9R14 - Baseline

```
$ make test TEST="micro:java.lang.ArraysSort.intSort" MICRO="VM_OPTIONS=-XX:UseAVX=2"

Benchmark            (size)  Mode  Cnt      Score      Error  Units
ArraysSort.intSort       10  avgt    3      0.035 ?    0.016  us/op
ArraysSort.intSort       25  avgt    3      0.091 ?    0.009  us/op
ArraysSort.intSort       50  avgt    3      0.245 ?    0.002  us/op
ArraysSort.intSort       75  avgt    3      0.412 ?    0.004  us/op
ArraysSort.intSort      100  avgt    3      0.531 ?    0.003  us/op
ArraysSort.intSort     1000  avgt    3      8.803 ?    0.609  us/op
ArraysSort.intSort    10000  avgt    3    254.413 ?  153.004  us/op
ArraysSort.intSort   100000  avgt    3   4485.811 ?   17.517  us/op
ArraysSort.intSort  1000000  avgt    3  56552.132 ? 3124.280  us/op
```

**Review by vnkozlov — APPROVED** [review]
My tier1-3,xcomp testing for v04 passed. I am integrating these changes.
Lets continue discussion about changes for AMD in  https://bugs.openjdk.org/browse/JDK-8317976.

**Comment by vamsi-parasa** [comment]
> My tier1-3,xcomp testing for v04 passed. I am integrating these changes. Lets continue discussion about changes for AMD in https://bugs.openjdk.org/browse/JDK-8317976.

Thank you, Vladimir!

File: make/modules/java.base/Lib.gmk
Comment (erikj79) [reply make/modules/java.base/Lib.gmk:230]: Ok, I missed that, never mind me then.

**Comment by PaulSandoz** [comment]
> At least on Saphire Rapids the [emulation suggested here](https://github.com/natmaurice/x86-simd-sort/commit/41d03b2d8f3b62a2ee6a3a97a8da7f193a407026) only imposes a 6% penalty for `intSort`, while also mitigating the performance issue on Zen 4.

Thanks for testing this out, this seems a good solution to explore. 

Is there is a way to `ifdef` switching on the CPU vendor? otherwise we could probably set a flag in the build script based of the CPU vendor, which may be more robust if different compilers are used.

**Comment by tarsa** [comment]
The answer for slow performance of AVX512 version of x86-simd-sort on Zen 4 is most probably explained in AMD manuals which could be found at: https://www.amd.com/en/search/documentation/hub.html#q=software%20optimization%20guide%20for%20the%20amd%20microarchitecture&f-amd_document_type=Software%20Optimization%20Guides

[Software Optimization Guide for the AMD Zen4 Microarchitecture](https://www.amd.com/content/dam/amd/en/documents/processor-tech-docs/software-optimization-guides/57647.zip) has following remark in "2.11.2 Code recommendations" chapter:
> Avoid the memory destination form of COMPRESS instructions.  These forms are implemented using microcode and achieve a lower store bandwidth than their register destination forms which use fastpath macro ops.

[Software Optimization Guide for the AMD Zen5 Microarchitecture](https://www.amd.com/content/dam/amd/en/documents/processor-tech-docs/software-optimization-guides/58455.zip) doesn't have any remark about COMPRESS instructions.

Could you add some code that disables the AVX512 version on Zen4, but keeps it enabled on Zen5 and future Zen architectures?

**Pull request closed without merging**
